Allow margins to collapse & refactor block toolbar

jasmussen opened the pull request:
The primary purpose of this PR is to allow margins to collapse. 

When two margins meet, and no borders or paddings are present, they collapse. I.e. a `p` with a 20px margin next to a similar `p` will have only 20px margin between them, not 40px, because the two siblings collapse to whichever margin is the largest. 

By allowing margins on blocks to collapse, we can make it easier to theme the editor to look like the frontend, by making margins behave the same.

This PR also refactors how the block toolbar is positioned. Because we use `position: sticky;`, the toolbar is part of the flow of the block itself. In this PR it uses `translate` to pull it _out of the flow_, meaning the block metrics do not have to account for the presence of the block toolbar or not. This is a nice refactor on its own, but it also makes the positioning more resilient and predictable.

## Testing

Because this is a rather large change to a visible part of the editor, it needs a good amount of testing. The gold standard is that **the editor in this branch should look no different than the editor in master**.

So please test on all breakpoints, test the frontend and backend, test with fixed toolbar and contextual toolbar:

- Test that the block toolbar that looks right on all blocks, including blocks like the spacer that do not have a block toolbar.
- Test all blocks they look right, margin-wise.
- Test in nested contexts blocks, test child blocks. 
- Test with editor styles. 

It's okay that there are a few regressions with existing editor styles, the benefits outweigh any fixes that need to be made.

Commit:
Allow collapsing margins & adjust block toolbar

This PR refactors how the block toolbar is positioned. Because we use sticky, the toolbar is part of the flow of the block itself. In this PR it uses translate to pull it out of the flow, making the positioning more resilient.

The primary purpose of this PR is to allow margins to collapse. When two margins meet, and no borders or paddings are present, they collapse. I.e. a `p` with a 20px margin next to a similar `p` will have only 20px margin between them, not 40px, because the two siblings collapse to whichever margin is the largest. By allowing margins on blocks to collapse, we can make it easier to theme the editor to look like the frontend.

jasmussen commented:
This has been rebased, and I feel like it's in solid shape for wider testing, and I think we should get it in 3.2. Aside from the code cleanliness and benefits this has for editor-themers, the way the block outlines and toolbar are painted now are more resilient. Because it's hard to explain, here are two GIFs. 

Here, I am adjusting the margins on the _block content itself_, in master, where margins do not collapse:

![without](https://user-images.githubusercontent.com/1204802/41714386-855bf826-7550-11e8-9f98-a5f0c3229c54.gif)

Note how the block toolbar does not correctly position itself according to the adjusted margins. 

Here's from the branch that allows collapsing margins:

![with](https://user-images.githubusercontent.com/1204802/41714420-9adf4d06-7550-11e8-89ab-e33d1821cdbd.gif)

Note how changing just one margin on the block content doesn't do anything — because they collapse so the remaining margin is still there. Also notice how when the other margin is changed, the outlines and block toolbar correctly position themselves.

aduth commented on editor/components/block-list/style.scss:
```diff
@@ -210,7 +209,8 @@
 	 */
-	.editor-block-list__block-edit {
+
+	 .editor-block-list__block-edit {
```
Mixed whitespace. Preceding space should be removed.

jasmussen replied:
Nice catch, thank you.

aduth reviewed (commented):
Block toolbar is very misplaced in IE11

![image](https://user-images.githubusercontent.com/1779930/41860621-15a91750-786d-11e8-9777-d0d3915bf305.png)

jasmussen commented:
Addressed feedback. Fixed the whitespace (nice catch, it shows up as a dot in VS code, I should've seen it but somehow missed it), and made it work in IE again. I hadn't properly rebased for Brandon's fix earlier, but now it works:

![edge](https://user-images.githubusercontent.com/1204802/41896960-47723c58-7926-11e8-94ab-6d47c371127a.png)

![ie](https://user-images.githubusercontent.com/1204802/41896962-4984b5e8-7926-11e8-9770-ff2d94534bef.PNG)

Noting also for others as well as myself, that `@supports( position: sticky ) { ... }` seems like a great way to target IE11 but not Edge, even for issues outside of position sticky. For example I might want to refactor the gallery hack that I made, to make things work in IE11, to use the above instead.

jasmussen commented:
Rebased, and squashed to keep future rebases simpler. In a `fixup`'ed commit, https://github.com/WordPress/gutenberg/pull/7365/files#diff-00ebf096f7c483426f2a78d70d9cbd22R1003, I included a fix for a horizontal scrollbar regression introduced in #7225.

jasmussen commented:
Unsure why Travis fails here. The tests pass.

tofumatt commented:
That has been intermittently failing a lot. I've restarted it. 😢

karmatosed reviewed (approved):
Design side approving and did some testing. I would like to have more testing just to be super sure.

jorgefilipecosta commented:
Hi @jasmussen, thank you for this changes. I confirm it solves a regression introduced during the refactoring to the block hover area.
I noticed a regression when clicking on the toolbar even if we move the mouse one pixel we trigger multi-selection. Before that was not the case. I'm not certain what is causing this problem maybe solve z-index change.
![jul-03-2018 16-09-54](https://user-images.githubusercontent.com/11271197/42228669-6de49604-7edc-11e8-9294-d9d01d1e7674.gif)

tofumatt commented:
FYI fixed the merge conflict here (via `git rebase`, so anyone working on this branch'll need to pull this new version).

tofumatt reviewed (commented):
This works for me from my testing, but I noticed the same regression @jorgefilipecosta did.

I found the cause… I'll push a commit fix for it.

tofumatt committed:
docs: Tweak comment in CSS

tofumatt committed:
fix: Multi-block selection on mouse movement after mouseDown

tofumatt commented:
I pushed a fix (https://github.com/WordPress/gutenberg/pull/7365/commits/cb8aae8de4c21993f8b4b879694425637fc6771e) for the regression with mouse movement on `mouseDown` @jorgefilipecosta mentioned and rebased this against master as there have been a few conflicts.

jasmussen committed:
Fix rebase issues, part 1

- Fixed toolbar position on mobile.
- Fixed toolbar position on desktop breakpoints
- Fixed appender regression
- Fixed issue with warning blocks having no hover or selected outlines.
- Fixed positioning of ellipsis on wide blocks.
- Fixed issue with full-wide gallery.

jasmussen commented on editor/components/block-list/block.js:
```diff
@@ -522,7 +523,7 @@ export class BlockListBlock extends Component {
 					data-block={ uid }
 				>
-
+					{ shouldShowContextualToolbar && <BlockContextualToolbar /> }
```
This move was initially the key to unlocking collapsing margins. So I assumed that aspect would break when you moved this back. 

However later on I added https://github.com/WordPress/gutenberg/pull/7365/commits/3c0afac08b5cdbac268793329ffed27e9674d604#diff-00ebf096f7c483426f2a78d70d9cbd22R904, which as it turns out is sufficient to enable collapsing margins, even without being relocated.

However all the in-between styles were written to assume the position of the nested toolbar where it was before this move, so there were a ton of visual regressions as a result of this move, like toolbars not being in the right place, and so on. But I'm fixing those now, and in the end I think it was good to move this back. Thanks for all the help.

jasmussen committed:
Fix rebase issues, part 2.

- Fixed a few typos and an issue with warnings. This will be revisited in a future patch also being worked on separately.
- Fixed position of toolbar on wide and fullwide.
- Fixed positioning of mover on wide and fullwide.
- Fixed issue with additional margins in columns block.
- Fixed issue with the invisible toolbar on the Column (singular) block breaking the layout. This also fixed an issue with the hover label and sibling inserter positioning on the column.

jasmussen committed:
Fix rebase issues, part 3.

- Fixed a regression in IE.
- Fixed the position of editor warning after the refactor.

jasmussen commented:
Okay, so I fixed a _ton_ of issues that propped up after the columns merge and the fix to the mouse movement regression. So many that I think this needs another deep dive into the code. Can you take a look, @tofumatt? Thanks. 

I left comments for each fix in each commit.

Once again tests pass for me with `npm run test` but Travis fails. Not sure why.

Also, in master and this branch, every image is broken for me, even after an `npm install`, I'm getting this error:

```
react-dom.min.82e21c65.js:110 TypeError: this.props.forwardedRef is not a function
    at Wrapper.handleRef (index.js:58)
    at cg (react-dom.min.82e21c65.js:110)
    at dg (react-dom.min.82e21c65.js:110)
    at eg (react-dom.min.82e21c65.js:113)
    at jc (react-dom.min.82e21c65.js:131)
    at gc (react-dom.min.82e21c65.js:127)
    at vb (react-dom.min.82e21c65.js:126)
    at ub (react-dom.min.82e21c65.js:126)
    at zd (react-dom.min.82e21c65.js:124)
    at ra (react-dom.min.82e21c65.js:123)
```

But given that's also in master, I'm guessing it's unrelated to this branch.

tofumatt commented on editor/components/block-list/style.scss:
```diff
@@ -121,15 +126,27 @@
+	.editor-default-block-appender__content,
+	.editor-block-list__block .editor-block-list__block-edit,
+	.editor-block-list__layout .editor-block-list__block .editor-block-list__block-edit,
+	.editor-block-list__layout .editor-default-block-appender .editor-default-block-appender__content { // Explicitly target nested blocks, as those need to override the preceeding rule.
```
Typo: "preceding". English is weird. I've pushed a fix 😄

tofumatt committed:
chore: Typo tweaks

tofumatt commented:
`npm install` works for me so I'm guessing it's local to your install, and that failing travis test is… frequent and happening a lot. We'll need to look into it. 😢

jasmussen commented:
To clarify, npm install is working fine for me. What isn't, is any image or embed. Try loading the demo content, I'm pretty sure you should see the same issue.

tofumatt commented:
As mentioned in Slack, this seems unrelated but a legit breakage: https://wordpress.slack.com/archives/C02QB2JS7/p1530703711000116

tofumatt reviewed (approved):
<del>I like this a lot more, it feels more like a real, empty canvas (and more like the old editor) with the tighter margins.</del> I have no idea what I'm on about and I think that's an unrelated change. I see what this is doing now. All good. More coffee needed.

I think we need to sort out that image bug but as mentioned it's not related to this PR and seems to be happening in master.

🚢 … but maybe after we fix master just so we don't merge breaking builds 😉

tofumatt commented:
As mentioned in Slack though: this is a great example of a PR that should include before/after screenshots. The differences are easy to spot visually but I found the lack of issue to make it unclear what I was testing for.

jasmussen commented:
Thanks a bunch for the review. I'll wait for master to behave and then merge, possibly rebase first.

jasmussen commented:
Rebased. Images work now, and the rest does too. 

How do we figure out if the travis failure is real? And otherwise, can we merge?

gziolo commented:
> How do we figure out if the travis failure is real? And otherwise, can we merge?

It would be very surprising if CSS changes would make any test fail. I retriggered another build.

gziolo commented on editor/components/block-list/style.scss:
```diff
@@ -801,42 +798,36 @@
 	text-align: left;
 	pointer-events: none;
-	height: $block-toolbar-height;
+	//height: $block-toolbar-height;
```
We should remove this line before merging.

jasmussen replied:
Fixed, thanks. I suppose this triggered another test? :)

jasmussen committed:
Remove comment.

gziolo committed:
Tests: Make adding blocks e2e test pass by refering the area below editor explicitly

aduth commented on test/e2e/specs/adding-blocks.test.js:
```diff
@@ -10,23 +10,9 @@ describe( 'adding blocks', () => {
 	it( 'Should insert content using the placeholder and the regular inserter', async () => {
 		// Click below editor to focus last field (block appender)
-		await clickBelow( await page.$( '.editor-default-block-appender' ) );
+		await page.click( '.editor-writing-flow__click-redirect' );
```
This refactor may have made the test less effective at covering the expected behavior: https://github.com/WordPress/gutenberg/pull/11168#pullrequestreview-169297243

jasmussen replied:
What's the best way to enhance this test again? CC: @gziolo who graciously helped me with this.

gziolo replied:
Ooops, I wasn't aware of that :(
